udp-native/src/main/java/org/ballerinalang/stdlib/udp/UdpListener.java

```diff
@@ -0,0 +1,101 @@
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.ballerinalang.stdlib.udp;
+
+import io.ballerina.runtime.api.Future;
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.socket.DatagramPacket;
+import io.netty.channel.socket.nio.NioDatagramChannel;
+import io.netty.handler.timeout.IdleStateHandler;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link UdpClient} creates the udp client and handles all the network operations.
+ */
+public class UdpListener {
+
+    private Channel channel;
+    private final EventLoopGroup group;
+    private final Bootstrap listenerBootstrap;
+
+    public UdpListener(InetSocketAddress localAddress, InetSocketAddress remoteAddress,
+                       EventLoopGroup group, Future callback, UdpService udpService) throws InterruptedException {
+        this.group = group;
+        listenerBootstrap = new Bootstrap();
+        listenerBootstrap.group(group)
+                .channel(NioDatagramChannel.class)
+                .handler(new ChannelInitializer<>() {
+                    @Override
+                    protected void initChannel(Channel ch) throws Exception {
+                        ch.pipeline().addLast(Constants.LISTENER_HANDLER, new UdpListenerHandler(udpService));
+                        ch.pipeline().addLast(Constants.READ_TIMEOUT_HANDLER,
+                                new IdleStateHandler(udpService.getTimeout(), 0, 0, TimeUnit.MILLISECONDS));
+                    }
+                });
+        if (remoteAddress != null) {
+            connect(remoteAddress, localAddress, callback);
+        } else {
+            channel = listenerBootstrap.bind(localAddress).sync().channel();
+            callback.complete(null);
+        }
+    }
+
+    public static void send(DatagramPacket datagram, Channel channel, Future callback) {
+        channel.writeAndFlush(datagram).addListener((ChannelFutureListener) future -> {
+            if (!future.isSuccess()) {
+                callback.complete(Utils.createSocketError("Failed to send data."));
+            } else {
+                callback.complete(null);
```

Comment:
Here we should set the error right instead of setting null.

Because sendBytes can return error.

check caller->sendBytes(data);

Reply:
yes sendBytes can return Error? , in write failure (.ie) !future.isSuccess() we are creating and returning the error on success we are returning null

Reply:
Oh missed it. In that case shall we remove the `!` and flip the code.

Reply:
fliped